pkg/modifycontroller/modify_status.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modifycontroller
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-csi/external-resizer/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// markControllerModifyVolumeStatus will mark ModifyVolumeStatus other than completed in the PVC
+func (ctrl *modifyController) markControllerModifyVolumeStatus(
+	pvc *v1.PersistentVolumeClaim,
+	modifyVolumeStatus v1.PersistentVolumeClaimModifyVolumeStatus,
+	err error) (*v1.PersistentVolumeClaim, error) {
+
+	newPVC := pvc.DeepCopy()
+	if newPVC.Status.ModifyVolumeStatus == nil {
+		newPVC.Status.ModifyVolumeStatus = &v1.ModifyVolumeStatus{}
+	}
+	newPVC.Status.ModifyVolumeStatus.Status = modifyVolumeStatus
+	// Update PVC's Condition to indicate modification
+	pvcCondition := v1.PersistentVolumeClaimCondition{
+		Type:               v1.PersistentVolumeClaimVolumeModifyingVolume,
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+	}
+	conditionMessage := ""
+	switch modifyVolumeStatus {
+	case v1.PersistentVolumeClaimModifyVolumeInProgress:
+		conditionMessage = "ModifyVolume operation in progress."
+		newPVC.Status.ModifyVolumeStatus.TargetVolumeAttributesClassName = *pvc.Spec.VolumeAttributesClassName
+	case v1.PersistentVolumeClaimModifyVolumeInfeasible:
+		conditionMessage = "ModifyVolume failed with error" + err.Error() + ". Waiting for retry."
+	}
+	pvcCondition.Message = conditionMessage
+	if modifyVolumeStatus != v1.PersistentVolumeClaimModifyVolumePending {
+		newPVC.Status.Conditions = util.MergeModifyVolumeConditionsOfPVC(newPVC.Status.Conditions,
+			[]v1.PersistentVolumeClaimCondition{pvcCondition})
+	}
+
+	updatedPVC, err := ctrl.patchClaim(pvc, newPVC, true)
+	if err != nil {
+		return pvc, fmt.Errorf("mark PVC %q as modify volume failed, errored with: %v", pvc.Name, err)
+	}
+	// Remove this PVC from the uncertain cache since the status is known now
+	if modifyVolumeStatus == v1.PersistentVolumeClaimModifyVolumeInfeasible {
+		ctrl.removePVCFromModifyVolumeUncertainCache(pvc)
+	}
+	return updatedPVC, nil
```

Comment:
So now that we have moved this code here, there is lot of duplication between this controller and resize controller. Can we reuse code related to updating pvc, pv in API server between these 2 controllers?

Reply:
The status update logic is quite different between resize and modify though.

Reply:
They do not share the same controller though, do we want to combine two separate controller's functions together?

Reply:
But their purpose, utility and code is the same. I was thinking, may be we can move these common functions to a shared util code or something.

Reply:
If we move them together at package import, we are importing functions for both?